react/react1/week1/todolist/src/components/ToDoList.js

```diff
@@ -0,0 +1,153 @@
+import { useState } from "react";
+import { useEffect } from "react";
+import ItemRow from "./ItemRow";
+import BlackBorder from "./BlackBorder";
+import Header from "./Header";
+const API_URL =
+  "https://gist.githubusercontent.com/benna100/391eee7a119b50bd2c5960ab51622532/raw";
+
+function ToDoList() {
+  const [todoList, setTodoList] = useState("");
+  const [count, setCount] = useState(0);
+  const [todoitem, setTodoitem] = useState("");
+  const [updateditem, setUpdateditem] = useState("");
+  const [startDate, setStartDate] = useState(new Date());
+  const currentDate = new Date();
+  const currentDateHours = currentDate.setHours(0, 0, 0, 0);
+
+  //FETCH
+  const fetchList = async () => {
+    const response = await fetch(API_URL);
+    const items = await response.json();
+    setTodoList(
+      items.map(({ id, description, deadline }) => ({
+        id: id,
+        description: description,
+        deadline: deadline,
+        done: false,
+        editing: false,
+      }))
+    );
+  };
+  useEffect(() => {
+    fetchList();
+  }, []);
+
+  //DELETE
+  function deleteItem(deleteId) {
+    setTodoList(todoList.filter(({ id }) => id !== deleteId));
+  }
+  //EDIT
+  function editItem(editId) {
+    const newState = todoList.map((obj) => {
+      if (obj.id === editId) {
+        return { ...obj, editing: !obj.editing };
+      }
+      return obj;
+    });
+
+    setTodoList(newState);
+  }
+  //UPDATE DESCRIPTION
+  function updateItem(updateId) {
+    if (updateditem.trim().length !== 0) {
+      const newState = todoList.map((obj) => {
+        if (obj.id === updateId) {
+          return { ...obj, description: updateditem, editing: !obj.editing };
+        }
+        return obj;
+      });
+
+      setTodoList(newState);
+      setUpdateditem("");
+    } else {
+      console.log("Updated item can't be empty!");
+    }
+  }
+  //CHANGE STATUS
+  function changeStatus(doneId) {
+    const newState = todoList.map((obj) => {
+      if (obj.id === doneId) {
+        return { ...obj, done: !obj.done };
+      }
+      return obj;
+    });
+
+    setTodoList(newState);
+  }
+
+  //ADD TODO ITEM
+  function addTodo() {
+    //CHECK IF INPUT OR DATE IS EMPTY, IF DEADLINE BEFORE CURRENT DAY
+    let inputDate = new Date(startDate);
+    let inputDateHours = inputDate.setHours(0, 0, 0, 0);
+    if (
+      todoitem.trim().length !== 0 &&
+      startDate.length !== 0 &&
+      inputDateHours >= currentDateHours
+    ) {
+      setTodoList(
+        todoList.concat([
+          {
+            id: todoList.length + 1,
+            description: todoitem,
+            deadline: startDate.toISOString().slice(0, 10),
+            done: false,
+          },
+        ])
+      );
+    } else if (todoitem.trim().length === 0) {
+      console.log("input can't be empty!");
```

Comment:
instead of console.log the user error find a way to show the error to the user as an error message instead like setError(x) and then show the error

Reply:
yes I was looking for a way, but seems a proper error setting is available with class components only https://reactjs.org/docs/error-boundaries.html

Reply:
What I would do is have a state like `const [inputError, setInputError] = useState("")` and then instead of doing `console.log("input can't be empty") ` then do `setInputError("input can't be empty")` and then have a place like when you submit the data then under the input field display the error like `<p>{inputError}</p>`. This way you are displaying to the user that they are not meeting the criteria provided for them.